sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/TableIdentifierParserSuite.scala

```diff
@@ -30,13 +64,22 @@ class TableIdentifierParserSuite extends SparkFunSuite {
     // Illegal names.
     intercept[ParseException](parseTableIdentifier(""))
     intercept[ParseException](parseTableIdentifier("d.q.g"))
+  }
 
+  test("table identifier - keywords") {
     // SQL Keywords.
-    val keywords = Seq("select", "from", "where", "left", "right")
+    val keywords = Seq("select", "from", "where") ++ hiveNonReservedRegression
     keywords.foreach { keyword =>
       intercept[ParseException](parseTableIdentifier(keyword))
       assert(TableIdentifier(keyword) === parseTableIdentifier(s"`$keyword`"))
       assert(TableIdentifier(keyword, Option("db")) === parseTableIdentifier(s"db.`$keyword`"))
     }
   }
+
+  test("table identifier - non reserved keywords") {
+    // Hive keywords are allowed.
+    hiveNonReservedKeyword.foreach { nonReserved =>
+      assert(TableIdentifier(nonReserved) === parseTableIdentifier(nonReserved))
+    }
+  }
```

Comment:
@gatorsmile thanks for filing the PR.

The current identifier parsing is probably as lenient as it gets. We really cannot add much more keywords without breaking the parsing process, for instance when `LEFT` is made nonReserved the parser will turn ```select * from a left join b on a.id = b.id``` into an `INNER` join because it interprets `LEFT` as the alias for table `a`. Union/Intersect/Except are also examples of this. I do like the idea of better error messages and I don't know if ANTLR has facilities for this. Lets do this in a different PR/JIRA if we want to investigate this.

We want to support names that start with an underscore. This is mainly because of  a Datasets of tuples; column names will all start with an underscore. Hive disallows underscores because `CharSetName`s start with an underscore.